spirv_reflect.c

```diff
@@ -3580,6 +3750,282 @@ static SpvReflectResult ParseExecutionModes(
+#if SPIRV_REFLECT_ENABLE_CONSTANT_EVALUATION
+#ifdef _MSC_VER
+#define SPIRV_REFLECT_FORCEINLINE __forceinline
+#else
+#define SPIRV_REFLECT_FORCEINLINE __attribute__((always_inline)) inline
+#endif
+
+static SpvReflectResult SPIRV_REFLECT_FORCEINLINE EvaluateResult_Impl(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectEvaluationMode evaluation_mode, SpvReflectPrvParser* p_parser);
+
+// only need node for evaluation.. Type information should not be freed before evaluating for now.
+// may move type info into instruction info later, but now it seems not needed...
+static SpvReflectResult EvaluateResult_Do(SpvReflectPrvEvaluationNode* p_node)
+{
+  return EvaluateResult_Impl(NULL, p_node, SPIRV_REFLECT_EVALUATION_MODE_NORMAL, NULL);
+}
+// 1 PASS parser needs to read specOp and check instruction size
+static SpvReflectResult EvaluateResult_1PASS(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectPrvParser* p_parser)
+{
+  return EvaluateResult_Impl(p_eval, p_node, SPIRV_REFLECT_EVALUATION_MODE_AST_1PASS, p_parser);
+}
+
+// 2 PASS needs parser to get id operands and literals
+static SpvReflectResult EvaluateResult_2PASS(SpvReflectEvaluation* p_eval, SpvReflectPrvEvaluationNode* p_node, SpvReflectPrvParser* p_parser)
+{
+  return EvaluateResult_Impl(p_eval, p_node, SPIRV_REFLECT_EVALUATION_MODE_AST_2PASS, p_parser);
+}
+
+#endif
+
+static SpvReflectResult ParseSpecializationConstants(SpvReflectPrvParser* p_parser, SpvReflectShaderModule* p_module)
+{
+  p_module->specialization_constant_count = 0;
+  p_module->specialization_constants = NULL;
+
+  uint32_t constant_instruction_num = 0;
+  for (size_t i = 0; i < p_parser->node_count; ++i) {
+    SpvReflectPrvNode* p_node = &(p_parser->nodes[i]);
+    // constant types need to be tracked after parser is dead...
+    if (IS_CONSTANT_OP(p_node->op)) {
+        constant_instruction_num++;
+    }
+    if (p_node->op == SpvOpSpecConstantTrue || p_node->op == SpvOpSpecConstantFalse || p_node->op == SpvOpSpecConstant) {
+      p_module->specialization_constant_count++;
+    }
+  }
+
+  if (p_module->specialization_constant_count == 0) {
+    return SPV_REFLECT_RESULT_SUCCESS;
+  }
+
+  p_module->specialization_constants = (SpvReflectSpecializationConstant*)calloc(p_module->specialization_constant_count, sizeof(SpvReflectSpecializationConstant));
+  if(IsNull(p_module->specialization_constants)){
+    return SPV_REFLECT_RESULT_ERROR_ALLOC_FAILED;
+  }
+
+  // deal with WorkGroupSize builtin
+  if (p_module->_internal->builtin_WorkGroupSize != INVALID_VALUE) {
+    SpvReflectPrvNode* p_node = FindNode(p_parser, p_module->_internal->builtin_WorkGroupSize);
+    for (uint32_t j = 0; j < p_module->entry_point_count; ++j) {
+      if (p_module->entry_points[j].spirv_execution_model == SpvExecutionModelKernel ||
+        p_module->entry_points[j].spirv_execution_model == SpvExecutionModelGLCompute) {
+        p_module->_internal->entry_point_mode_flags[j] = 4;
```

Comment:
What is this code block doing? Why is it overwriting the x value of the local size with the OpID that WorkGroupSize is held in?

Reply:
When the code is written, specialization constants for workgroup size (aka `local_size_id_n`), as per spirv 1.5 specification, the compiler have two options for generating code. One is using `OpExecutionModeId`, and referencing resultid of three constant Ops. The other is using `WorkGroupSize` builtin, which takes a uvec3 argument resultid and overwrites the `OpExecutionMode` Op. When the code is written, `glslc` and other compilers are relying on the second (builtin) mode for generating code in glsl. The second mode used to be deprecated in spv1.5,  but later it seems the deprecation plan was moved to 1.6, if I remember correctly.

Reference:
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html

![Screenshot_20240229-074453](https://github.com/KhronosGroup/SPIRV-Reflect/assets/34389517/0674249f-1e76-42d6-bf99-1a733c6d134f)

![Screenshot_20240229-074419](https://github.com/KhronosGroup/SPIRV-Reflect/assets/34389517/51433a15-6a02-4f78-bd84-33ddaccf7ec0)

![Screenshot_20240229-074447](https://github.com/KhronosGroup/SPIRV-Reflect/assets/34389517/2242b382-3e57-45e3-b407-a8d25587cf56)

Reply:
For a reference of spv asm disassembled from glslc generated code that uses the builtin, see the sample test code included in pr:

https://github.com/shangjiaxuan/SPIRV-Reflect/blob/spec-constant/tests/spec_constant/test_orig.spv.dis

Reply:
Maybe the version of this PR I'm using is out of date, but in my case if a GLSL shader is using a spec constant, and is using the gl_WorkGroup variable, but the spec constant is **not** use to set the local_size, this code block still gets used and you end up with the incorrect value for local_size.

Reply:
> Maybe the version of this PR I'm using is out of date, but in my case if a GLSL shader is using a spec constant, and is using the gl_WorkGroup variable, but the spec constant is **not** use to set the local_size, this code block still gets used and you end up with the incorrect value for local_size.

Please give me sample glsl code and compiler options to compile it. Or you can give me the disassembled code. This can help clarify your problem and help me locate it.

Reply:
You should check out the entry mode flag that can be queried with `spvReflectGetEntryModeFlag`. In order to support specialized localsize, modes are added. The flag should have been in the localsize struct, but that would break abi and was requested to be moved to private member.

Reply:
The x member is overwritten to represent the uvec3 resultid if in builtin `WorkGroupSize` mode, and xyz are ovewritten as three resultid in `ExecutionModeId`.

Reply:
The rationale is that there might be specconstant arithmetic, whether scalar or vector. `WorkGroupSize` takes precedence over `ExecutionMode`, and the arguments used to be (1, 1, 1). It gives (16, 32, 1) default values with `WorkGroupSize` mainly for compatibility.

Reply:
hmm not sure I follow. In this case I want to get the values (16, 32, 1), since that is the actual local_size values. Where will those values reside when it's in this mode?

Reply:
You should use `spvReflectEvaluateResult` to evaluate the value of localsize.x resultid, for the current specialization constants in the evaluator (default values if you don't set them) to get the current uvec3.

Reply:
Ok, I can see how that'll result in me getter the correct answer. However, specialization constant's aren't used for the workgroup size in this example. The are hardcoded constants. It seems to me it should only enter into this mode if a specialization constant is actually used to set the workgroup size.

Reply:
It also baffles me why the compiler generates the builtin instruction. The older version compiler didn't do this if I remember correctly.

Reply:
I'm currently compiled again GLSLang 302a663a38, so it's not the most up to date version.
Another comment, I would think the reflection should put this result_id somewhere else, and try to resolve the correct local size in to the local_size variable. Re-using 'x' like this is confusing, and in general the user of reflection just wants the results, so asking them to use spvReflectEvaluateResult seems like an un-needed step.

A comment about the ABI (not directed at @shangjiaxuan), I don't understand why it would be requested that this tool which is only 2 files, should have a locked ABI. It's not completed enough to start hampering the interface with a locked ABI at this point.